examples/fluid.py

```diff
@@ -0,0 +1,233 @@
+    for i, j in vf:
+        vl = sample_v(vf, i - 1, j)[0]
+        vr = sample_v(vf, i + 1, j)[0]
+        vt = sample_v(vf, i, j + 1)[1]
+        vb = sample_v(vf, i, j - 1)[1]
+        div_vels[i, j] = ((vr - vl) + (vt - vb)) * half_rdx
+
+
+p_alpha = -dx * dx
+
+
+@ti.kernel
+def pressure_jacobi(pf: ti.template(), new_pf: ti.template()):
+    for i, j in pf:
+        pl = sample_p(pf, i - 1, j)
+        pr = sample_p(pf, i + 1, j)
+        pt = sample_p(pf, i, j + 1)
+        pb = sample_p(pf, i, j - 1)
+        div = div_vels[i, j]
+        new_pf[i, j] = (pl + pr + pt + pb + p_alpha * div) * 0.25
+
+
+@ti.kernel
+def subtract_gradient(new_vf: ti.template(), pf: ti.template()):
+    # Make |new_vf| divergence free
+    for i, j in new_vf:
+        pl = sample_p(pf, i - 1, j)
+        pr = sample_p(pf, i + 1, j)
+        pt = sample_p(pf, i, j + 1)
+        pb = sample_p(pf, i, j - 1)
+        v = new_vf[i, j]
+        v -= half_rdx * ti.Vector([pr - pl, pt - pb])
+        v *= v_decay
+        new_vf[i, j] = v
+
+
+@ti.kernel
+def fill_color(new_vf: ti.template()):
+    for i, j in new_vf:
+        v = new_vf[i, j]
+        color_buffer[i, j] = ti.Vector([ti.abs(v[0]), ti.abs(v[1]), 0.25])
+
+
+def reset():
+    velocities.fill(ti.Vector([0.0, 0.0]))
+    pressures.fill(0.0)
+    color_buffer.fill(ti.Vector([0.0, 0.0, 0.0]))
+
+
+def step(mouse_data):
+    global velocities, new_velocities, pressures, new_pressures
+    advect(velocities, new_velocities)
+    if enable_diffusion:
+        for _ in range(v_jacobi_iters):
+            diffusion_jacobi(velocities, new_velocities)
+            velocities, new_velocities = new_velocities, velocities
+    apply_mouse_impulse(new_velocities, mouse_data)
+    divergence(new_velocities)
+    pressures.fill(0.0)
+    for _ in range(p_jacobi_iters):
+        pressure_jacobi(pressures, new_pressures)
+        pressures, new_pressures = new_pressures, pressures
+    subtract_gradient(new_velocities, pressures)
+    fill_color(new_velocities)
+    velocities, new_velocities = new_velocities, velocities
+
+
+def vec2_npf32(m):
+    return np.array([m[0], m[1]], dtype=np.float32)
+
+
+class MouseDataGen(object):
+    def __init__(self):
+        self.prev_mouse = None
+
+    def __call__(self, gui):
+        # [0:2]: normalized delta direction
+        # [2:4]: current mouse xy
+        mouse_data = np.array([0] * 4, dtype=np.float32)
+        if gui.is_pressed(ti.GUI.LMB):
+            mxy = vec2_npf32(gui.get_cursor_pos()) * res
+            if self.prev_mouse is None:
+                self.prev_mouse = mxy
+            else:
+                mdir = mxy - self.prev_mouse
+                mdir = mdir / (np.linalg.norm(mdir) + 1e-5)
+                mouse_data[0], mouse_data[1] = mdir[0], mdir[1]
+                mouse_data[2], mouse_data[3] = mxy[0], mxy[1]
+                self.prev_mouse = mxy
+        else:
+            self.prev_mouse = None
+        return mouse_data
```

Comment:
Nice work! Mouse drag diff is useful for such program, I'm considering add this into GUI system.

Reply:
Right, that will be very nice to have!

Reply:
That sounds like a good idea - let's merge this in for now and upgrade the GUI system in a later PR. Thanks!